File: refractor-agent/scripts/evaluate.py
#!/usr/bin/env python3
"""Evaluate refractor-agent against the golden dataset (eval/golden.yaml).

Two modes:
  match  — match-layer only: use each case's pre-recorded ``rec`` and run the
           matching logic. Offline, no VLM needed (embeddings still used).
  live   — full pipeline: recognize the card images with the VLM first, then
           match. Needs VLM_* env vars and image files.

Prints a metrics table; also writes metrics.json / errors.jsonl / confusion.json
under the output dir. ``--sweep`` scans thresholds for calibration.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import yaml

from match import best_type, query_text, series_name
from refract_store import (
    BaseStore,
    Embedder,
    cosine,
    create_store,
    load_env,
    LOCAL_THRESHOLD,
    REMOTE_THRESHOLD,
)
from vlm import recognize

DEFAULT_GOLDEN = Path(__file__).resolve().parent.parent / "eval" / "golden.yaml"
DEFAULT_OUT = Path(__file__).resolve().parent.parent / "eval" / "out"
DICT_DIR = Path(__file__).resolve().parent.parent / "dicts"
SWEEP_STEPS = [round(0.50 + 0.05 * i, 2) for i in range(10)]  # 0.50..0.95


def load_cases(golden: Path, images_root: Path) -> list[dict]:
    with golden.open(encoding="utf-8") as fh:
        doc = yaml.safe_load(fh)
    cases = []
    for c in doc.get("cases", []):
        c["images_root"] = images_root
        cases.append(c)
    return cases


def prepare(cases: list[dict], mode: str, embedder: Embedder) -> list[dict]:
    """Recognize (live) and embed the query once per case; reuse for every threshold."""
    prepared = []
    for case in cases:
        item = {"case": case}
        try:
            if mode == "live":
                imgs = case_images(case)
                if not imgs:
                    item["error"] = "missing images"
                else:
                    item["rec"] = recognize(imgs)
            else:  # match
                rec = case.get("rec")
                if not rec:
                    item["error"] = "case has no rec (match mode)"
                else:
                    item["rec"] = rec
            if "rec" in item and item["rec"].get("pattern") != "平卡":
                item["qv"] = embedder.embed([query_text(item["rec"])])[0]
        except Exception as exc:  # noqa: BLE001 - a bad case must not kill the run
            item["error"] = str(exc)
            item.pop("rec", None)
        prepared.append(item)
    return prepared


def run_match_prepared(item: dict, rows: list[dict], store: BaseStore,
                       threshold: float) -> dict:
    """Match using a precomputed query vector (``qv``): global type match +
    series naming lookup (same logic as scripts/match.py)."""
    rec = item.get("rec")
    if rec is None:
        return {"error": item.get("error", "no rec")}
    if rec.get("pattern") == "平卡":
        return {"matched": False, "refraction": None, "needsReview": False}

    scores = [(cosine(item["qv"], r["vector"]), r) for r in rows]
    typ = best_type(scores, threshold)
    if typ is None:
        return {"matched": False, "needsReview": True}

    naming = series_name(rec, typ["pattern"], typ["color"], DICT_DIR)
    if naming is None:
        return {**typ, "refraction": None, "needsReview": True}
    return {**typ, **naming, "needsReview": False}


def effective_threshold(args_threshold: float | None, embedder: Embedder) -> float:
    if args_threshold is not None:
        return args_threshold
    return REMOTE_THRESHOLD if embedder.remote else LOCAL_THRESHOLD


def case_images(case: dict) -> list[Path]:
    paths = []
    for key in ("front", "back"):
        rel = case.get(key)
        if rel:
            p = case["images_root"] / rel
            if p.is_file():
                paths.append(p)
    return paths


def aggregate(prepared: list[dict], rows: list[dict], store: BaseStore,
              threshold: float) -> tuple[dict, list[dict], dict]:
    """Run matching over prepared cases at one threshold and aggregate metrics."""
    details: list[dict] = []
    confusion: dict[str, dict[str, int]] = {}
    det_c = term_c = pattern_c = color_c = series_c = 0
    non_plain = predicted_terms = review = total = 0

    for item in prepared:
        total += 1
        case = item["case"]
        expected = case.get("expected", {})
        pred = run_match_prepared(item, rows, store, threshold)
        if "error" in pred:
            details.append({"id": case["id"], "error": pred["error"], "stage": "recognize"})
            continue
        exp_refr = expected.get("refraction")
        pred_refr = pred.get("refraction")

        det_ok = (exp_refr is None) == (pred_refr is None)
        det_c += int(det_ok)

        exp_key = exp_refr or "(plain)"
        pred_key = pred_refr or "(none)"
        confusion.setdefault(exp_key, {})
        confusion[exp_key][pred_key] = confusion[exp_key].get(pred_key, 0) + 1

        fields = {}
        if pred_refr is not None:
            predicted_terms += 1
        rec = item.get("rec") or {}
        if exp_refr is not None:
            non_plain += 1
            term_ok = pred_refr == exp_refr
            term_c += int(term_ok)
            fields["term"] = term_ok
            if expected.get("pattern") is not None:
                pattern_ok = pred.get("pattern") == expected["pattern"]
                pattern_c += int(pattern_ok)
                fields["pattern"] = pattern_ok
            if expected.get("color") is not None:
                color_ok = pred.get("color") == expected["color"]
                color_c += int(color_ok)
                fields["color"] = color_ok
            # series accuracy: does the RECOGNITION self-determine the right
            # brand x series? (that decides which naming table to look up)
            if expected.get("brand") is not None and expected.get("series") is not None:
                series_ok = (rec.get("brand"), rec.get("series")) == (
                    expected["brand"], expected["series"])
                series_c += int(series_ok)
                fields["series"] = series_ok
        if pred.get("needsReview"):
            review += 1
        fields["review"] = bool(pred.get("needsReview"))

        details.append({
            "id": case["id"],
            "expected": expected,
            "rec": item.get("rec"),
            "predicted": pred,
            "correct": fields,
        })

    def pct(n: int, d: int) -> float:
        return round(n / d, 4) if d else None

    metrics = {
        "total": total,
        "det_acc": pct(det_c, total),
        "term_acc": pct(term_c, non_plain),
        "pattern_acc": pct(pattern_c, non_plain),
        "color_acc": pct(color_c, non_plain),
        "series_acc": pct(series_c, non_plain),
        "review_rate": pct(review, total),
        "precision": pct(term_c, predicted_terms),
        "recall": pct(term_c, non_plain),
        "non_plain": non_plain,
        "predicted_terms": predicted_terms,
    }
    return metrics, details, confusion


def print_metrics(m: dict) -> None:
    def fmt(v) -> str:
        return "n/a" if v is None else f"{v:.2%}"

    print("=" * 46)
    print(f"  total cases      : {m['total']}")
    for k in ("det_acc", "term_acc", "pattern_acc", "color_acc", "series_acc"):
        print(f"  {k:<18}: {fmt(m.get(k))}")
    print(f"  review_rate      : {fmt(m.get('review_rate'))}")
    print(f"  precision        : {fmt(m.get('precision'))}")
    print(f"  recall           : {fmt(m.get('recall'))}")
    print("=" * 46)


def write_outputs(out_dir: Path, metrics: dict, details: list[dict], confusion: dict) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)
    (out_dir / "metrics.json").write_text(
        json.dumps(metrics, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    with (out_dir / "errors.jsonl").open("w", encoding="utf-8") as fh:
        for d in details:
            fh.write(json.dumps(d, ensure_ascii=False) + "\n")
    (out_dir / "confusion.json").write_text(
        json.dumps(confusion, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--mode", choices=["match", "live"], default="match")
    ap.add_argument("--golden", default=str(DEFAULT_GOLDEN))
    ap.add_argument("--images-root", default=None, help="override golden images_root")
    ap.add_argument("--db", default=str(Path(__file__).parent.parent / "db" / "refractors.lance"))
    ap.add_argument("--threshold", type=float, default=None)
    ap.add_argument("--sweep", action="store_true", help="scan thresholds (match mode)")
    ap.add_argument("--out", default=str(DEFAULT_OUT))
    args = ap.parse_args()

    load_env()
    embedder = Embedder()
    store = create_store(args.db)
    rows = store.rows()
    if not rows:
        raise SystemExit("empty store; run scripts/embed.py first")
    golden = Path(args.golden)
    images_root = Path(args.images_root) if args.images_root else golden.parent
    cases = load_cases(golden, images_root)
    if not cases:
        print(f"no cases in {golden}", file=sys.stderr)
        return 1

    # recognize + embed every query once; thresholds only re-rank
    prepared = prepare(cases, args.mode, embedder)

    if args.sweep:
        print(f"{'thr':>5} {'term_acc':>9} {'review_rate':>11} {'recall':>7}")
        for t in SWEEP_STEPS:
            m, _, _ = aggregate(prepared, rows, store, t)
            ta = m["term_acc"] if m["term_acc"] is not None else float("nan")
            rr = m["review_rate"] if m["review_rate"] is not None else float("nan")
            rc = m["recall"] if m["recall"] is not None else float("nan")
            print(f"{t:>5.2f} {ta:>9.2%} {rr:>11.2%} {rc:>7.2%}")
        return 0

    threshold = effective_threshold(args.threshold, embedder)
    metrics, details, confusion = aggregate(prepared, rows, store, threshold)
    print_metrics(metrics)
    write_outputs(Path(args.out), metrics, details, confusion)
    print(f"detail rows -> {Path(args.out) / 'errors.jsonl'}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: refractor-agent/node/src/evaluate.ts
/** Evaluate the refractor matching logic against the golden dataset.
 *
 * TypeScript port of `scripts/evaluate.py` (match mode; live mode needs the
 * VLM adapter, which is not ported yet). Prints a metrics table and writes
 * metrics.json / errors.jsonl / confusion.json under the output dir. The
 * Python version is the behavioral contract — the golden double-run compares
 * metrics, detail rows, and confusion output for exact equality.
 */

import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { parse } from "yaml";

import { AGENT_ROOT, loadEnv } from "./env.ts";
import { bestType, queryText, seriesName } from "./match.ts";
import {
  BaseStore,
  Embedder,
  LOCAL_THRESHOLD,
  REMOTE_THRESHOLD,
  createStore,
  ensureSynced,
  type TypeHit,
} from "./store.ts";
import { rebuildStore } from "./embed.ts";

export interface GoldenCase {
  id: string;
  front?: string;
  back?: string;
  expected?: Record<string, unknown>;
  rec?: Record<string, unknown>;
}

export interface PreparedItem {
  case: GoldenCase;
  rec?: Record<string, unknown>;
  qv?: number[];
  hit?: [number, TypeHit] | null;
  error?: string;
}

export interface MatchOutput {
  matched?: boolean;
  refraction?: string | null;
  name_en?: string | null;
  pattern?: string;
  color?: string;
  matchScore?: number;
  needsReview?: boolean;
  error?: string;
}

export interface DetailRow {
  id: string;
  error?: string;
  stage?: string;
  expected?: Record<string, unknown>;
  rec?: Record<string, unknown>;
  predicted?: MatchOutput;
  correct?: Record<string, boolean>;
}

export type Confusion = Record<string, Record<string, number>>;

export interface Metrics {
  total: number;
  det_acc: number | null;
  term_acc: number | null;
  pattern_acc: number | null;
  color_acc: number | null;
  series_acc: number | null;
  review_rate: number | null;
  precision: number | null;
  recall: number | null;
  non_plain: number;
  predicted_terms: number;
}

/** Python `round(value, digits)` — half-even on the decimal representation. */
export function roundHalfEven(value: number, digits: number): number {
  const factor = 10 ** digits;
  const scaled = value * factor;
  const floor = Math.floor(scaled);
  const diff = scaled - floor;
  let rounded: number;
  if (diff > 0.5) rounded = floor + 1;
  else if (diff < 0.5) rounded = floor;
  else rounded = floor % 2 === 0 ? floor : floor + 1;
  return rounded / factor;
}

export async function loadCases(goldenPath: string): Promise<GoldenCase[]> {
  const doc = parse(await readFile(goldenPath, "utf8")) as { cases?: GoldenCase[] };
  return doc.cases ?? [];
}

/** Recognize (live) and embed the query once per case; reused per threshold. */
export async function prepareMatch(
  cases: readonly GoldenCase[],
  embedder: Embedder,
): Promise<PreparedItem[]> {
  const prepared: PreparedItem[] = [];
  for (const caseItem of cases) {
    const item: PreparedItem = { case: caseItem };
    try {
      const rec = caseItem.rec;
      if (!rec) {
        item.error = "case has no rec (match mode)";
      } else {
        item.rec = rec;
        if (rec.pattern !== "平卡") {
          item.qv = (await embedder.embed([queryText(rec)]))[0];
        }
      }
    } catch (error) {
      item.error = error instanceof Error ? error.message : String(error);
      delete item.rec;
    }
    prepared.push(item);
  }
  return prepared;
}

/** Match a prepared case: top1 hit + series naming (same logic as match). */
export async function runMatchPrepared(
  item: PreparedItem,
  store: BaseStore,
  threshold: number,
): Promise<MatchOutput> {
  const rec = item.rec;
  if (rec === undefined) return { error: item.error ?? "no rec" };
  if (rec.pattern === "平卡") {
    return { matched: false, refraction: null, needsReview: false };
  }

  const typ = bestType(item.hit ?? null, threshold);
  if (typ === null) {
    return { matched: false, needsReview: true };
  }

  const naming = await seriesName(rec, typ.pattern, typ.color, store);
  if (naming === null) {
    return { ...typ, refraction: null, needsReview: true };
  }
  return { ...typ, ...naming, needsReview: false };
}

/** Run matching over prepared cases at one threshold and aggregate metrics. */
export async function aggregate(
  prepared: readonly PreparedItem[],
  store: BaseStore,
  threshold: number,
): Promise<{ metrics: Metrics; details: DetailRow[]; confusion: Confusion }> {
  const details: DetailRow[] = [];
  const confusion: Confusion = {};
  let detCount = 0;
  let termCount = 0;
  let patternCount = 0;
  let colorCount = 0;
  let seriesCount = 0;
  let nonPlain = 0;
  let predictedTerms = 0;
  let reviewCount = 0;
  let total = 0;

  for (const item of prepared) {
    total += 1;
    const expected = item.case.expected ?? {};
    const pred = await runMatchPrepared(item, store, threshold);
    if (pred.error !== undefined) {
      details.push({ id: item.case.id, error: pred.error, stage: "recognize" });
      continue;
    }
    const expRefr = typeof expected.refraction === "string" ? expected.refraction : null;
    const predRefr = typeof pred.refraction === "string" ? pred.refraction : null;

    const detOk = (expRefr === null) === (predRefr === null);
    detCount += detOk ? 1 : 0;

    const expKey = expRefr ?? "(plain)";
    const predKey = predRefr ?? "(none)";
    const row = (confusion[expKey] ??= {});
    row[predKey] = (row[predKey] ?? 0) + 1;

    const fields: Record<string, boolean> = {};
    if (predRefr !== null) predictedTerms += 1;
    const rec = item.rec ?? {};
    if (expRefr !== null) {
      nonPlain += 1;
      const termOk = predRefr === expRefr;
      termCount += termOk ? 1 : 0;
      fields.term = termOk;
      if (expected.pattern != null) {
        const patternOk = pred.pattern === expected.pattern;
        patternCount += patternOk ? 1 : 0;
        fields.pattern = patternOk;
      }
      if (expected.color != null) {
        const colorOk = pred.color === expected.color;
        colorCount += colorOk ? 1 : 0;
        fields.color = colorOk;
      }
      // series accuracy: does the RECOGNITION self-determine the right
      // brand x series? (that decides which naming table to look up)
      if (expected.brand != null && expected.series != null) {
        const seriesOk =
          rec.brand === expected.brand && rec.series === expected.series;
        seriesCount += seriesOk ? 1 : 0;
        fields.series = seriesOk;
      }
    }
    if (pred.needsReview === true) reviewCount += 1;
    fields.review = pred.needsReview === true;

    details.push({
      id: item.case.id,
      expected,
      rec,
      predicted: pred,
      correct: fields,
    });
  }

  const pct = (count: number, denominator: number): number | null =>
    denominator > 0 ? roundHalfEven(count / denominator, 4) : null;

  const metrics: Metrics = {
    total,
    det_acc: pct(detCount, total),
    term_acc: pct(termCount, nonPlain),
    pattern_acc: pct(patternCount, nonPlain),
    color_acc: pct(colorCount, nonPlain),
    series_acc: pct(seriesCount, nonPlain),
    review_rate: pct(reviewCount, total),
    precision: pct(termCount, predictedTerms),
    recall: pct(termCount, nonPlain),
    non_plain: nonPlain,
    predicted_terms: predictedTerms,
  };
  return { metrics, details, confusion };
}

export interface RunMatchOptions {
  golden?: string;
  db?: string;
  threshold?: number;
  out?: string;
  dictDir?: string;
}

export interface MatchRunResult {
  metrics: Metrics;
  details: DetailRow[];
  confusion: Confusion;
}

/** Full match-mode evaluation flow (evaluate.py main, match branch). */
export async function runMatchEvaluation(options: RunMatchOptions = {}): Promise<MatchRunResult> {
  loadEnv();
  const dictDir = options.dictDir ?? join(AGENT_ROOT, "dicts");
  const dbPath = options.db ?? join(AGENT_ROOT, "db", "refractors.lance");
  const goldenPath = options.golden ?? join(AGENT_ROOT, "eval", "golden.yaml");

  // Unlike the Python version, the self-healing rebuild is bound to the same
  // store path being evaluated (the Python rebuild always targets the default
  // store, even when --db points elsewhere).
  const sync = await ensureSynced(dictDir, () => rebuildStore(dictDir, dbPath));
  if (sync !== "synced") {
    console.error(`[sync] vector store ${sync}`);
  }

  const embedder = new Embedder();
  const store = await createStore(dbPath);
  const cases = await loadCases(goldenPath);
  if (cases.length === 0) {
    throw new Error(`no cases in ${goldenPath}`);
  }

  const prepared = await prepareMatch(cases, embedder);
  for (const item of prepared) {
    if (item.qv !== undefined) {
      item.hit = await store.top1(item.qv);
    }
  }

  const threshold =
    options.threshold ?? (embedder.remote ? REMOTE_THRESHOLD : LOCAL_THRESHOLD);
  const { metrics, details, confusion } = await aggregate(prepared, store, threshold);

  const outDir = options.out ?? join(AGENT_ROOT, "eval", "out");
  await mkdir(outDir, { recursive: true });
  await writeFile(
    join(outDir, "metrics.json"),
    JSON.stringify(metrics, null, 2) + "\n",
    "utf8",
  );
  await writeFile(
    join(outDir, "errors.jsonl"),
    details.map((detail) => JSON.stringify(detail)).join("\n") + "\n",
    "utf8",
  );
  await writeFile(
    join(outDir, "confusion.json"),
    JSON.stringify(confusion, null, 2) + "\n",
    "utf8",
  );
  return { metrics, details, confusion };
}

function printMetrics(metrics: Metrics): void {
  const fmt = (value: number | null): string =>
    value === null ? "n/a" : `${(value * 100).toFixed(2)}%`;
  console.log("=".repeat(46));
  console.log(`  total cases      : ${metrics.total}`);
  for (const key of ["det_acc", "term_acc", "pattern_acc", "color_acc", "series_acc"] as const) {
    console.log(`  ${key.padEnd(18)}: ${fmt(metrics[key])}`);
  }
  console.log(`  review_rate      : ${fmt(metrics.review_rate)}`);
  console.log(`  precision        : ${fmt(metrics.precision)}`);
  console.log(`  recall           : ${fmt(metrics.recall)}`);
  console.log("=".repeat(46));
}

async function main(): Promise<number> {
  const argv = process.argv.slice(2);
  let mode = "match";
  let golden: string | undefined;
  let db: string | undefined;
  let threshold: number | undefined;
  let out: string | undefined;
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--mode") mode = argv[++i] as string;
    else if (argv[i] === "--golden") golden = argv[++i] as string;
    else if (argv[i] === "--db") db = argv[++i] as string;
    else if (argv[i] === "--threshold") threshold = Number(argv[++i]);
    else if (argv[i] === "--out") out = argv[++i] as string;
    else if (argv[i] === "--images-root" || argv[i] === "--sweep") {
      // accepted for CLI parity; match mode ignores them (sweep support pending)
    }
  }
  if (mode !== "match") {
    console.error("live mode is not ported yet (requires the VLM adapter)");
    return 2;
  }

  const { metrics } = await runMatchEvaluation({ golden, db, threshold, out });
  printMetrics(metrics);
  return 0;
}

if (process.argv[1] !== undefined && import.meta.url === new URL(`file://${process.argv[1]}`).href) {
  main().then(
    (code) => process.exit(code),
    (error) => {
      console.error(error instanceof Error ? error.message : error);
      process.exit(1);
    },
  );
}
